content/blog/2022/2022-07-19-Nodejs-Logging-Novice-to-Expert.md

```diff
@@ -0,0 +1,567 @@
+
+But first, lets see the Winston library in action. Returning to our `app.js`. Here we can replace all `console` statements with our newly created `logger`:
+
+_change code in app.js to this._
+
+```javascript
+const express = require("express");
+const logger = require("./logger");
+const app = express();
+
+app.get("/", (req, res, next) => {
+  logger.log("debug", "Hello, Winston!");
+  logger.debug("The is the home '/' route.");
+  res.status(200).send("Logging Hello World..");
+});
+
+app.get("/event", (req, res, next) => {
+  try {
+    throw new Error("Not User!");
+  } catch (error) {
+    logger.error("Events Error: Unauthenticated user");
+    res.status(500).send("Error!");
+  }
+});
+
+app.listen(3000, () => {
+  logger.info("Server Listenning On Port 3000");
+});
+```
+
+Run `node app.js` to start the server.
+
+When we access the above routes via the paths `/` and `/event`, we get logs in JSON format.
+
+```bash
+{"level":"info","message":"Server Listenning On Port 3000"}
+{"level":"debug","message":"Hello, Winston!"}
+{"level":"debug","message":"The is the home '/' route."}
+{"level":"error","message":"unauthenticated user failed"}
+```
+
+## Winston Method Options
+
+As seen above the `winston.createLogger()` method gives us a number of options to help format and transport our logs.
+
+Let us visit each option and examine the properties and features they offers.
+
+## **Winston Level**
+
+Log level is the piece of information in our code that indicates the importance of a specific log message. Using appropriate log levels is one of the best practices for application logging.
+
+Winston's logging levels follow the severity ordering specified by [RFC5424](https://tools.ietf.org/html/rfc5424) and npm logging levels. Custom levels can also be added.
+
+In the npm logging level, the severity of all levels is prioritized from the most important to least important from 0 to 6 (highest to lowest).
+
+- 0: error
+- 1: warn
+- 2: info
+- 3: http
+- 4: verbose
+- 5: debug
+- 6: silly
+
+Using the appropriate Log level to log, allows you to consume and route logging output based on message importance.
+
+_where:_
+
+- **Error**: is a serious problem or failure, that halts current activity but leaves the application in a recoverable state with no effect on other operations. The application can continue working
+- **Warn**: A non-blocking warning about an unusual system exception. These logs provide context for a possible error. it logs warning signs that should be investigated.
+- **Info**: This denotes major events and informative messages about the application's current state. Useful For tracking the flow of the application.
+- **HTTP**: This logs out HTTP request-related messages. HTTP transactions ranging from the host, path, response, requests, etc.
+- **Verbose**: Records detailed messages that may contain sensitive information. Verbose was [renamed](https://github.com/aspnet/Announcements/issues/124) to Trace.
```

Comment:
How is an ASP.NET log level relevant to a Node.js log level?

Reply:
Okay i will update this